### ISSUE TITLE
Cannot get InteractiveBrowserCredential() to cache credentials

### ISSUE THREAD

**Issue by rfernand2** [Azure/azure-sdk-for-python#23721]
- **Package Name**:  azure-identity
- **Package Version**: 1.7.0
- **Operating System**: Windows 11
- **Python Version**: 3.8.0

**Describe the bug**
Using InteractiveBrowserCredential() with enable_persistent_cache=True does not seem to cache the credentials over subsequent runs (even for as little as 30 seconds).  When I also specify the authenication_record parameter, it still prompts for credentials in the browse on each run.

**To Reproduce**

Run the following python program:
```
# auth_with_cach_test.py: test out ability to get credentials without having to authenticate each time
import os
from azure.identity import InteractiveBrowserCredential, AuthenticationRecord, TokenCachePersistenceOptions

fn = "cred_cache.txt"
deserialized_record = None

found_cc = os.path.exists(fn)
if found_cc:
    with open(fn, "rt") as infile:
        cred_json = infile.read()
        deserialized_record = AuthenticationRecord.deserialize(cred_json)
        print("record read from:", fn)

credential = InteractiveBrowserCredential(enable_persistent_cache=True, authenication_record=deserialized_record)
record = credential.authenticate()

if not found_cc:
    # serialize to our cache
    cred_json = record.serialize()
    with open(fn, "wt") as outfile:
        outfile.write(cred_json)

    print("record written to:", fn)
```

**Expected behavior**
The first time the program is run, I expect it to prompt for credentials in the browser.  The second time the program is run, I expect it to run to completion without prompting for credentials in the browser, but it always prompts.

**Screenshots**

**Additional context**
See original issue: #9744.  Note: I am looking for the simplest way to get an "auth" to pass to the Azure Workspace.get() method, since it seems to require one with for certain Singularity workspaces.  I would prefer, in the following order:
   - to just log-in using my Azure portal login cached creds - is there a way to do this yet?
   - to have Azure cache my credentials itself (in the cloud, or on my local computer) - is there a way to do this yet?
   - to be able to cache some bit of information that I can use to get the "auth" without having to make the user authenticate again
 The above test program represents the suggested code from #9744, as I interpret it.  Did I miss something?

**Comment by xiangyan99** [comment]
Thanks for the feedback. Here is sample code to use token cache:

```py
cred = InteractiveBrowserCredential(tenant_id=tenant_id, cache_persistence_options=TokenCachePersistenceOptions())
token = cred.get_token(YOUR_SCOPE)
auth_record = cred.authenticate()
cred1 = InteractiveBrowserCredential(tenant_id=tenant_id,
                                         cache_persistence_options=TokenCachePersistenceOptions(),
                                         authentication_record=auth_record)
token1 = cred1.get_token(YOUR_SCOPE)
```

Does this work for you?

**Comment by rfernand2** [comment]
No, that still requires browser authentication every time I run it.  BTW, is this new code supposed to have **Azure** or the **user** do the credential persistence?

Here is the complete version of your new code that I tested:

```
# auth_with_cache_test2.py: test ability to get Azure creds without always authenticating
import os
from azure.identity import InteractiveBrowserCredential, AuthenticationRecord, TokenCachePersistenceOptions

scopes = ""
tenant_id = "72f988bf-86f1-41af-91ab-2d7cd011db47"

cred = InteractiveBrowserCredential(tenant_id=tenant_id, cache_persistence_options=TokenCachePersistenceOptions())
token = cred.get_token(scopes)
auth_record = cred.authenticate()
cred1 = InteractiveBrowserCredential(tenant_id=tenant_id,
                                         cache_persistence_options=TokenCachePersistenceOptions(),
                                         authentication_record=auth_record)
token1 = cred1.get_token(scopes)
print("token1=", token1)

```

**Comment by xiangyan99** [comment]
I doubt "" is a valid scope.

Could you try something like

"https://management.core.windows.net/.default"?

We expect to see a browser pop up in the first request

token = cred.get_token(scopes)

And on the second one

token1 = cred1.get_token(scopes)

It should come from the cache so there should no browser pop up.

**Comment by rfernand2** [comment]
Hi @xiangyan99 - using the above URL for scopes, it still authenticates in the browser, on each run of the test app.  Are you able to repro the problem with my program?

**Comment by xiangyan99** [comment]
No. I cannot repro it. 

Here is my app:

```py
from azure.identity import InteractiveBrowserCredential, TokenCachePersistenceOptions

def main():
    tenant_id = 'YOUR_TENANT_ID'
    cred = InteractiveBrowserCredential(tenant_id=tenant_id, cache_persistence_options=TokenCachePersistenceOptions())
    token = cred.get_token("https://management.core.windows.net/.default")
    auth_record = cred.authenticate()
    cred1 = InteractiveBrowserCredential(tenant_id=tenant_id,
                                         cache_persistence_options=TokenCachePersistenceOptions(),
                                         authentication_record=auth_record)
    token1 = cred1.get_token("https://management.core.windows.net/.default")
    print("pass")

if __name__ == '__main__':
    main()
```

The browser only pops up once.

**Comment by rfernand2** [comment]
I just ran your program 4 times - each time, the browser got focus for authentication.  One thing that might be different about my environment - I don't run "az" (Azure CLI).   Is this design somehow dependent on it being installed?

**Comment by rfernand2** [comment]
Here is my program:

# sing_test2.py: test out ability to create a workspace object for a singularity workspace
import os
from azureml.core.authentication import TokenAuthentication
from azure.identity import InteractiveBrowserCredential, AuthenticationRecord, TokenCachePersistenceOptions

fn = "cred_cache.txt"
tenant_id = None
deserialized_record = None
scope = https://management.core.windows.net/.default
tenant_id = "72f988bf-86f1-41af-91ab-2d7cd011db47"

if os.path.exists(fn):
    # read serialized cred from disk
    with open(fn, "rt") as infile:
        cred_json = infile.read()
        deserialized_record = AuthenticationRecord.deserialize(cred_json)
        #tenant_id = deserialized_record.tenant_id

cpo = TokenCachePersistenceOptions()
credential = InteractiveBrowserCredential(tenant_id=tenant_id,
    #enable_persistent_cache=True,
    cache_persistence_options=cpo,
    authenication_record=deserialized_record)

if not os.path.exists(fn):
    # serialize to our cred to disk
    record = credential.authenticate()
    cred_json = record.serialize()

    with open(fn, "wt") as outfile:
        outfile.write(cred_json)

# prove the credential is valid
token = credential.get_token(scope)
print(token)

**Comment by rfernand2** [comment]
Thanks to help from @xiangyan99, the below, updated program now works correctly (Azure authentication in browser is only done on first run):

```
# creds_cache_test.py: test out ability to use cached creds to avoid authenticating on every run
import os
from azure.identity import InteractiveBrowserCredential, AuthenticationRecord, TokenCachePersistenceOptions 

# actual credential caching is handling by azure, but our app must supply the authentication record 
# which identifies the tentant_id, username, etc.
fn = "cred_cache.txt"
deserialized_record = None

if os.path.exists(fn):
    # read serialized authentication_record from local file
    with open(fn, "rt") as infile:
        cred_json = infile.read()
        deserialized_record = AuthenticationRecord.deserialize(cred_json)

cpo = TokenCachePersistenceOptions()
credential = InteractiveBrowserCredential(cache_persistence_options=cpo,
    authentication_record=deserialized_record)

if not os.path.exists(fn):
    # serialize authentication_record to local file
    record = credential.authenticate()
    cred_json = record.serialize()

    with open(fn, "wt") as outfile:
        outfile.write(cred_json)

# prove the credential is valid
scope = "https://management.core.windows.net/.default"
token = credential.get_token(scope)
print(token)
```

**Comment by jgentil** [comment]
Why is this closed?  Shouldn't InteractiveBrowserCredential try to load from the cache first?

**Comment by xiangyan99** [comment]
Yes. And this is the code snippet how to achieve it.

https://github.com/Azure/azure-sdk-for-python/issues/23721#issuecomment-1083539872

**Comment by jgentil** [comment]
Maybe I don't understand the design decision here.  When I look at SharedTokenCacheCredential, if I don't pass an authentication_record parameter, it figures it out for me.  Why isn't this the case for InteractiveBrowserCredential?  

My work-around for this was to use a chained credential of (SharedTokenCacheCredential, InteractiveBrowserCredential) so that they loaded.  But it's silly that InteractiveBrowserCredential can't figure out how to use the cache without an authentication_record when SharedTokenCacheCredential can.

**Comment by xiangyan99** [comment]
If you don't mind, could you share your code snippets how you use those credentials?

My guess is you did not specify tenant_id for them. If it is not specified, SharedTokenCacheCredential gives a default one but InteractiveBrowserCredential cannot give a default one hence it prompts you to specify.
